### ISSUE TITLE
fidelity of floating point emulation

### ISSUE THREAD

**Issue by dat20gh** [NationalSecurityAgency/ghidra#2414]
I am writing a sleigh spec for a "new" processor. To check my work I am executing the pcode and verifying it against
a "known good" model (the actual hardware, for the problem below).

The issue I have run into is that floating point division (FLOAT_DIV) is generating minor errors compared to the actual hardware.

Interestingly, the problem reproduces on x86. So, I expect the issue is generic for FLOAT_DIV and would apply to all architectures. The problem can be reproduced with a "tiny" rom image, and a python script... 

bash$ printf '%b%b%b%b' "\xf3" "\x0f" "\x5e" "\xc1" > /tmp/x86.rom

Loading this into ghidra at 0x0 as x86:LE:64:default should give "DIVSS XMM0,XMM1" as disassembly.
This can then be executed in conjunction with the small python script (below):

./analyzeHeadless /tmp x86 -noanalysis -import /tmp/x86.rom -loader BinaryLoader -loader-baseAddr 0x00000000  -loader-blockName myblock -processor x86:LE:64:default  -postScript /tmp/x86.py

In the output, it will print:
0x40bb8000

the correct value, as executed on x86, (and my target hardware) is: 0x40bb8001 <- note ends in 1... 

Can someone familiar with the inner workings of the float code comment on whether this can be fixed by setting a rounding mode or some other "global" control? I am using ghidra_9.1.2_PUBLIC and jdk-11.0.8

cat > /tmp/x86.py

```from ghidra.app.emulator import EmulatorHelper

def main():
    emuHelper = EmulatorHelper(currentProgram)
    emuHelper.writeRegister("XMM0", 0x42580000)
    emuHelper.writeRegister("XMM1", 0x411374bc)
    emuHelper.writeRegister("RIP", 0)
    emuHelper.writeRegister("RSP", 0)
    emuHelper.step(monitor)
    print "{:#010x}".format(emuHelper.readRegister("XMM0"))

main()
```

**Comment by jpleasu** [comment]
The emulator floating point arithmetic is done by the class `ghidra.pcode.floatformat.FloatFormat`.  The method `opDiv` converts each target value to native Java `double`, does the division, then returns `getEncoding` of the result.  The bug is in `getEncoding`.

The mantissa of Java's double precision result is truncated from 52 to 23 bits, it ought to be rounded.

see [here](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/pcode/floatformat/FloatFormat.java#L373)

**Comment by mumbel** [comment]
@jpleasu  May be of some interest/related to this issue, but I played around a bit with pcodetest and FP operations and it was not going well and couldn't get reliable results. (#1479)

**Comment by jpleasu** [comment]
@mumbel  Thanks for the pointer -- you're right, they're related.  `EmulatorHelper` is used by `ProcessorEmulatorTestAdapter`, the base class for the pcode tests.  Your convert tests make for a really good goal.

**Comment by dat20gh** [comment]
Thanks for your comments! It's taken me a while to circle back to this...

@jpleasu the pointer you gave into the code helped me greatly...

I built a new version of ghidra with this one line patch:

```
git diff
diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/pcode/floatformat/FloatFormat.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/pcode/floatformat/FloatFormat.java
index 2b97ac6..b1e7c33 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/pcode/floatformat/FloatFormat.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/pcode/floatformat/FloatFormat.java
@@ -370,6 +370,8 @@ public class FloatFormat {
                if (exp != 0 && jbitimplied)
                        signif <<= 1; // Cut of top bit for normal case only (which should be 1)
 
+               signif = signif + ((1L << (64 - frac_size)) >> 1);
+
                long res = (signif >>> (64 - frac_size)) << frac_pos;
                res |= (long) exp << exp_pos;
                return setSign(res, data.sign);

```

and that proved to be correct for 100 or so tests cases that I have in a practical use case...

Caveat emptor though :-) I cannot claim that the test cases I have constitute coverage... and I'm not truly confident in the fix as I don't understand the ramifications of:

```
                if (exp != 0 && jbitimplied)
                        signif <<= 1; // Cut of top bit for normal case only (which should be 1)
```

Anyway, I hope posting this comment may help others. If I get some time, I will look into this some more.

**Comment by jpleasu** [comment]
Sorry for the lack of updates - I'm almost done with a pretty thorough set of fixes for Ghidra's emulation of floating point.  It should go in for review this week.

Ghidra's emulation of floating point is incorrect in two ways -- the first, which you ran into, is the rounding.  IEEE754 prescribes rounding to the nearest even for binary floating point values, but Ghidra's emulator truncates.  Ghidra doesn't need to round when 
emulating doubles (since native doubles are used), but anything else and Ghidra will be wrong.  The second issue is incomplete support for "subnormal" numbers.

I test against pseudo-random values to catch rounding errors and for the many corner cases I include +- zero, +- infinity, nan, +-min, max subnormal, max subnormal +- min, and +-max.

**Comment by dat20gh** [comment]
@jpleasu thanks for the update.

I have found an issue that I think is related... This time in int2float. I have an instruction in my sleigh code which (paraphrasing is):

``` rd = int2float(ra) ```

where ra and rd are 32b. When ra is set to (0x80000000 | 0x12) the result in rd is "ce800000" -- on the actual hardware, the result is cf000000 [some c on x86 also produces cf000000]. As you can see the mantissa is 0, but the exponent field (bits [30..23]) is wrong 157 (0x9d) vs the correct 158 (0x9e)... Which (I'm guessing) means the exponent hasn't been correctly adjusted somewhere.

I'm wondering if this is on a common code path with the float issues you are chasing down(??) and is just another aspect of this issue.

FWIW: the int2float conversion of 0x12 is correct --- when I wrote the test harnesss, I threw in the or'ing of 0x80000000 to create an additional test and this raised the issue!

**Comment by jpleasu** [comment]
Yes, it's a common path and the new emulator produces the correct result, I tested with:

```java
OpBehaviorFloatInt2Float op = new OpBehaviorFloatInt2Float();
long result = op.evaluateUnary(4, 4, 0x80000000 | 0x12);
System.err.printf("%x\n", result);
// produces:
// cf000000
```
`evaluateUnary` calls to `FloatFormat#opInt2Float` which calls `getEncoding`, where erroneous truncation happens.

**Comment by dat20gh** [comment]
@jpleasu Thanks for explaining the issue with int2float !

**Comment by ghidra1** [comment]
Corrected for 9.2.3 release
